matrix-sdk/src/main/java/org/matrix/androidsdk/login/AutoDiscovery.kt

```diff
@@ -0,0 +1,191 @@
+         if the client is inclined to do so. Failure can take place instead if no good user experience for this is possible at this point.
+         */
+        PROMPT,
+        /*
+         Stop the current auto-discovery mechanism. If no more auto-discovery mechanisms are available,
+         then the client may use other methods of determining the required parameters, such as prompting the user, or using default values.
+         */
+        IGNORE,
+        /*
+            Inform the user that auto-discovery failed due to invalid/empty data and PROMPT for the parameter.
+         */
+        FAIL_PROMPT,
+        /*
+         Inform the user that auto-discovery did not return any usable URLs. Do not continue further with the current login process.
+         At this point, valid data was obtained, but no homeserver is available to serve the client.
+         No further guess should be attempted and the user should make a conscientious decision what to do next.
+         */
+        FAIL_ERROR,
+
+    }
+
+
+    fun findClientConfig(domain: String, callback: ApiCallback<DiscoveredClientConfig>) {
+
+        discoveryRestClient.getWellKnown(domain, object : ApiCallback<WellKnown> {
+            override fun onSuccess(wellKnown: WellKnown) {
+                if (wellKnown.homeServer?.baseURL.isNullOrBlank()) {
+                    callback.onSuccess(DiscoveredClientConfig(Action.FAIL_PROMPT))
+                } else {
+                    val baseURLString = wellKnown.homeServer!!.baseURL!!
+                    if (!isValidURL(baseURLString)) {
+                        callback.onSuccess(DiscoveredClientConfig(Action.FAIL_ERROR))
+                        return
+                    }
+                    //Check that HS is a real one
+                    validateHomeServerAndProceed(wellKnown, callback)
+                }
+            }
+
+            override fun onUnexpectedError(e: Exception) {
+                callback.onUnexpectedError(e)
+            }
+
+            override fun onNetworkError(e: Exception) {
+                if (e is HttpException) {
+                    when (e.httpError.httpCode) {
+                        HttpsURLConnection.HTTP_NOT_FOUND -> callback.onSuccess(DiscoveredClientConfig(Action.IGNORE))
+                        else -> callback.onSuccess(DiscoveredClientConfig(Action.FAIL_PROMPT))
+                    }
+                } else if (e is MalformedJsonException || e is EOFException) {
+                    callback.onSuccess(DiscoveredClientConfig(Action.FAIL_PROMPT))
+                } else {
+                    callback.onNetworkError(e);
+                }
+            }
+
+            override fun onMatrixError(e: MatrixError) {
+                callback.onMatrixError(e)
+            }
+
+        })
+    }
+
+    private fun validateHomeServerAndProceed(wellKnown: WellKnown, callback: ApiCallback<DiscoveredClientConfig>) {
```

Comment:
Look like a SuccessCallback is enough here

Reply:
I don't think i can do that?  param must be an ApiCallback